manual/node_pattern.md

```diff
@@ -0,0 +1,270 @@
+# Node Pattern
+
+Node pattern is a small language to help find specific nodes in
+the Abstract Syntax Tree using a simple string.
+
+It reminds the simplicity of regular expressions but used to find specific
+nodes of ruby code.
+
+## Simple match
+
+A simple integer like `1` represents `(int 1)` in the AST.
+
+```
+$ ruby-parse -e '1'
+(int 1)
+```
+
+- `int` will match exactly the node
+- `(int 1)` will match precisely the node
+
+## `_` for any value
+
+`_` will check if there's something present in the specific position, no matter the
+value:
+
+- `(int _)` will match any number
+- `(int _ _)` will not match because `int` types have just one child that
+  contains the value.
+
+The compiler translates the pattern `(int _)` into something like:
+
+```ruby
+node.int_type? && node.children.size == 1
+```
+
+while `(int _ _)` will be translated to something like:
+
+```ruby
+node.int_type? && node.children.size == 2
+```
+
+## `...` for something else
+
+While `_` limits the size, `...` make it open for more children.
+
+And the example `(int ...)` will be translated into something like:
+
+```ruby
+node.int_type? && node.children.size >= 1
+```
+
+And it matches with any number as well.
+
+It's useful when you want to check some variable internal nodes but with a
+final with the same results. For example, let's use a classic `something.save`.
+
+We can also have `person.save` or `person.address.save`, and we want to match
+both. So, let's check how it looks like in the AST:
+
+```
+$ ruby-parse -e 'person.save'
+(send (send nil :person) :save)
+```
+
+The first case can be addressed with an expression like:
+
+```
+(send (send nil? _) :save)
+```
+
+But if it contains `.address` in the middle:
+
+```
+$ruby-parse -e 'person.address.save'
+(send
+  (send
+    (send nil :person) :address) :save)
+```
+
+The expression will not match, and it can be `a.b.c.d.save`.
+
+In such cases you can use `...` to match both cases:
+
+```
+(send ... :save)
+```
+
+## `{}` for any `<expression>`
+
+Lets make it a bit more complex and introduce floats:
+
+```
+$ ruby-parse -e '1'
+(int 1)
+$ ruby-parse -e '1.0'
+(float 1.0)
+```
+
+- `({int float} _)` - int or float types, no matter the value
```

Comment:
This may be simplified to just `{int float}` I suppose.

I see {a b} as `a || b`.